marshmallow/schema.py

```diff
@@ -706,10 +701,32 @@ def _update_fields(self, obj=None, many=False):
         else:
             field_names = self.set_class(self.declared_fields.keys())
 
+        declared_fields = field_names
+        invalid_fields = self.set_class()
+        if self.only is not None:
+            # Return only fields specified in only option
+            only = self.set_class(self.only)
+            field_names = only
+            invalid_only = only - declared_fields
+            invalid_fields |= invalid_only
+
         # If "exclude" option or param is specified, remove those fields
         excludes = set(self.opts.exclude) | set(self.exclude)
         if excludes:
             field_names = field_names - excludes
+            invalid_excludes = excludes - declared_fields
+            invalid_fields |= invalid_excludes
+
+        if invalid_fields:
+            invalid_names = ['"{}"'.format(name) for name in invalid_fields]
+            invalid_names = utils.iterable_to_string(invalid_names)
```

Comment:
(Pylint nazi inside)

Reusing variable name is considered bad practice (R0204 (redefined-variable-type)).

You could append _list or _str to one of them, or just go for the one-liner equivalent.